Documentation/ApiOverview/DependencyInjection/Index.rst

```diff
@@ -12,59 +16,466 @@ Dependency injection
+   namespace MyVendor\MyExtension\Controller;
+
+   use MyVendor\MyExtension\Repository\UserRepository;
+
+   class UserController
    {
-       $this->foo = $foo;
-       $this->bar = $bar;
+      protected UserRepository $userRepository;
+
+      public function __construct(UserRepository $userRepository)
+      {
+         $this->userRepository = $userRepository;
+      }
    }
 
-By default all API services shipped by the TYPO3 Core system extensions are available for dependency
-injection.
+((This is straight and easy DI: The controller class is configured to be autowired
+as Symfony DI service container entry (more on that below), Symfony container finds a
+dependency to :php:`UserRepository` when scanning :php:`__construct()` and then knows that
+:php:`UserController` needs an instance of the repository service when the controller
+service is instantiated.)) ← ToDo: Improve understandability
 
-.. note::
 
-   The following document aims at explaining dependency injection as used in TYPO3.
-   For those new to the concepts and principles behind dependency injection, please
-   read more general introductions on the topic first - for example at
-   `PHP The Right Way <https://phptherightway.com/#dependency_injection>`_ and
-   at `Fabien Potenciers blog <http://fabien.potencier.org/what-is-dependency-injection.html>`_.
+.. _Method-injection:
 
-.. contents::
-   :depth: 3
+Method injection
+----------------
 
-.. index::
-   pair: Dependency injection; Extensions
-   pair: Dependency injection; Services
-   File; EXT:{extkey}/Configuration/Services.yaml
+A second way to get services injected is by using :php:`inject*()` methods:
 
-When to use Dependency Injection in TYPO3
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. code-block:: php
 
-Class dependencies to services should be injected via constructor injection or
-setter methods. Where possible, Symfony dependency injection
-should be used for all cases where DI is required.
+   namespace MyVendor\MyExtension\Controller;
+
+   use MyVendor\MyExtension\Repository\UserRepository;
+
+   class UserController
+   {
+      protected ?UserRepository $userRepository = null;
+
+      public function injectUserRepository(UserRepository $userRepository)
+      {
+         $this->userRepository = $userRepository;
+      }
+   }
+
+This ends up with the exactly the same result as above: The controller instance has
+an object of type :php:`UserRepository` in class property :php:`$userRepository`.
+This inject* way came from extbase and TYPO3 core implemented it in addition to
+the default Symfony constructor injection. Why did we do that, you may ask? Both
+strategies have subtle differences: First, when using inject* methods, the type
+hinted class property needs to be nullable, otherwise PHP >= 7.4 throws a warning
+since the instance is not set during :php:`__construct()`. But that's just a
+language detail. More important is an abstraction scenario. Consider this case:
```

Comment:
```suggestion
since the instance is not set during :php:`__construct()`. But that's just an
implementation detail. More important is an abstraction scenario. Consider this case:
```